### PR TITLE
Add .NET Core 3/.NET Standard 2.1 support.

### PR DESCRIPTION
**Pull request by PJB3005**
Unlike #70 this does not remove any code, because the relevant APIs have been added in .NET Standard 2.1/.NET Core 3.0. Obviously this does mean that it will not work on .NET Core 2/.NET Standard 2.0, but it's not that big of a loss since .NET Framework 4.5 is still explicitly supported. 

Debug assembly generation is disabled on platforms outside of Framework since the necessary API isn't exposed (and probably won't be for a very long time).

### CODE REVIEW

File: Test/Test.csproj [hunk 1]
```diff
@@ -1,80 +1,13 @@
...
-    <Compile Include="Tester.cs" />
-    <Compile Include="MessageTest.cs" />
-  </ItemGroup>
   <ItemGroup>
-    <ProjectReference Include="..\NetSerializer\NetSerializer.csproj">
-      <Project>{85A11D07-8D18-42D5-ACCF-EF9744EFE825}</Project>
-      <Name>NetSerializer</Name>
-    </ProjectReference>
+    <PackageReference Include="protobuf-net" Version="3.0.0-alpha.43" />
+    <PackageReference Include="Mono.Options" Version="5.3.0.1" />
```
Comment (tomba) [inline Test/Test.csproj]: Update the package dependencies separately. They shouldn't be part of this patch.

File: NetSerializer.sln.DotSettings.user [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+	<s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=PublicFields/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="AaBb"&gt;&lt;ExtraRule Prefix="" Suffix="" Style="aaBb_aaBb" /&gt;&lt;/Policy&gt;</s:String>
+	<s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=Property/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="AaBb" /&gt;</s:String>
+	<s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=Interfaces/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="I" Suffix="" Style="AaBb" /&gt;</s:String>
+	<s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=StaticReadonly/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="s_" Suffix="" Style="aaBb" /&gt;</s:String>
+	<s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=EnumMember/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="AaBb" /&gt;</s:String>
+	<s:Boolean x:Key="/Default/CodeStyle/Naming/CSharpAutoNaming/IsNamingAutoDetectionCompleted/@EntryValue">True</s:Boolean></wpf:ResourceDictionary>
```
Comment (tomba) [inline NetSerializer.sln.DotSettings.user]: User settings shouldn't be part of the project.

File: Test/Test.csproj [hunk 3]
```diff
@@ -3,8 +3,11 @@
     <OutputType>Exe</OutputType>
     <TargetFrameworks>net45;netcoreapp3.0</TargetFrameworks>
   </PropertyGroup>
+
+  <Import Project="../build.props" />
+
   <ItemGroup>
-    <PackageReference Include="protobuf-net" Version="3.0.0-alpha.43" />
+    <PackageReference Include="protobuf-net" Version="2.4.0" />
```
Comment (tomba) [inline Test/Test.csproj]: Don't go back and forth in your patches. If you do a mistake in an earlier patch, fix that patch, don't fix it in a separate patch in the same series.

And that goes for this whole patch too. "Fix build on Unix", but why is it broken? Because of the first patch? Then the first patch should be fixed.

File: NetSerializer.sln.DotSettings.user (hunk 2 above)
Comment (PJB3005) [reply NetSerializer.sln.DotSettings.user]: My bad, didn't mean to commit this.

File: Test/Test.csproj (hunk 1 above)
Comment (PJB3005) [reply Test/Test.csproj]: Yeah the problem is that both of the packages were too old to run on .NET Core so I had to update them. I'll make a separate PR though.

Comment (tomba) [reply Test/Test.csproj]: No need for a separate PR, just have the change in a separate commit.

File: NetSerializer/NetSerializer.csproj [hunk 4]
```diff
@@ -1,68 +1,25 @@
...
-    <Compile Include="TypeDictionary.cs" />
-    <Compile Include="Helpers.cs" />
-    <Compile Include="ITypeSerializer.cs" />
-    <Compile Include="Serializer.cs" />
-    <Compile Include="Primitives.cs" />
-    <Compile Include="TypeSerializers\NoOpSerializer.cs" />
-    <Compile Include="TypeSerializers\ObjectSerializer.cs" />
-    <Compile Include="TypeSerializers\PrimitivesSerializer.cs" />
-    <Compile Include="Properties\AssemblyInfo.cs" />
+    <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0-preview.2">
```
Comment (tomba) [inline NetSerializer/NetSerializer.csproj]: Hmm, what is this reference? A ref to "preview.2" doesn't sound good.

**Comment by tomba** [comment]
I think the patches look ok as such, but I don't understand what the limitations introduced are. E.g. in Ubuntu 18, I have mono installed from Ubuntu packages and dotnet installed from snap. But trying nuget with these patches gives:

`
$ nuget restore
MSBuild auto-detection: using msbuild version '15.0' from '/usr/lib/mono/msbuild/15.0/bin'.
/usr/lib/mono/msbuild/Current/bin/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(137,5): error NETSDK1045: The current .NET SDK does not support targeting .NET Core 3.0.  Either target .NET Core 2.1 or lower, or use a version of the .NET SDK that supports .NET Core 3.0. [/home/tomba/tmp/netserializer/Test/Test.csproj]
/usr/lib/mono/msbuild/Current/bin/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(137,5): error NETSDK1045: The current .NET SDK does not support targeting .NET Core 3.0.  Either target .NET Core 2.1 or lower, or use a version of the .NET SDK that supports .NET Core 3.0. [/home/tomba/tmp/netserializer/PrimitiveTest/PrimitiveTest.csproj]
  It was not possible to find any installed dotnet SDKs
  Did you mean to run dotnet SDK commands? Please install dotnet SDK from:
      https://aka.ms/dotnet-download
`

I understand that I may not get .NET Core 3 support without downloading the very latest packages, but plain .Net framework should still work, with the old tools, no?

File: NetSerializer/NetSerializer.csproj (hunk 4 above)
Comment (PJB3005) [reply NetSerializer/NetSerializer.csproj]: It's to allow the .NET Core SDK to build the project targeting .NET Framework on mac/Linux, because they can't use regular Framework (you can also make it rely on Mono being installed but this is more elegant and less hacky).

It's preview because .NET Core 3 is also still preview.

**Comment by PJB3005** [comment]
Normally you should be able to specify `/p:TargetFramework=net45` to make it not try to build the .NET Standard/Core targets so it will build on Mono, but I couldn't get `nuget` to accept that meaning package restore was still broken.

I made the build only support .NET Framework when using Mono to compile.

**Comment by tomba** [comment]
Alright, works for me.

I'm pretty out of touch about today's .Net development, but as the old way works for me, and the patches look fine, I think I'll merge this. I'll wait a day or two in the hopes that someone else gives a thumbs up too.

**Comment by PJB3005** [comment]
Would it be possible to update the NuGet package? We're depending on it and it's blocking us from cleanly upgrading our projects.

**Comment by tomba** [comment]
Updated
